### ISSUE TITLE
Resampled image is cropped

### ISSUE THREAD

**Issue by ctorti** [SimpleITK/SimpleITK#1358]
I'm working with 3D DICOM images and 3D binary label images and encountering issues when it comes to resampling.  At first glance it seemed that my issue might be related to #1160, but unlike the issue posted there, the structure I am trying to resample is nowhere near a boundary - hence it is not being cropped by the extent of the new gridspace.

I should note that I previously attempted this in [SimpleElastix](https://github.com/SuperElastix/SimpleElastix/issues/419) and encountered the same issue.  I have now re-created the effect using SimpleITK.

Rather than repeat the text that I already posted in the SimpleElastix GitHub issues I thought I'd take a more general approach here.  I'll start by creating a sitk image with the same size of the "Source" image of a cube centered within the image:

```
import numpy as np
import SimpleITK as sitk

SrcR, SrcC, SrcF = SrcIm.GetSize() # (512, 512, 192)

# Create a 3D mask of 1s of width w along x and y, and height h along z, centered in the image:
w = 100
h = 10

imin = (SrcR - w)//2
imax = (SrcR + w)//2

jmin = (SrcC - w)//2
jmax = (SrcC + w)//2

kmin = (SrcF - h)//2
kmax = (SrcF + h)//2

mask = np.zeros((SrcF, SrcC, SrcR))

mask[kmin:kmax, jmin:jmax, imin:imax] = 1

CubeIm = sitk.GetImageFromArray(mask)

CubeIm.SetOrigin(SrcIm.GetOrigin())
CubeIm.SetDirection(SrcIm.GetDirection())
CubeIm.SetSpacing(SrcIm.GetSpacing())
```

Following is the resulting 3D cube:

![SrcCube](https://user-images.githubusercontent.com/26715524/112358907-a84e8c80-8cc8-11eb-816c-90ff36a5ab29.jpg)

Next I register `SrcIm` to `TrgIm` to get the registered image and final transform:

EDIT 25/03/21:  In the original code that follows, I had interchangeably used `SrcIm` and `MovIm`, and `TrgIm` and `FixIm`.  I have now kept to `SrcIm` and `TrgIm`. 

```
InitialTx = sitk.CenteredTransformInitializer(TrgIm, SrcIm, 
                                              sitk.AffineTransform(TrgIm.GetDimension()), 
                                              sitk.CenteredTransformInitializerFilter.MOMENTS)

RegMethod = sitk.ImageRegistrationMethod()

""" Similarity metric settings. """
RegMethod.SetMetricAsMattesMutualInformation(numberOfHistogramBins=50)
RegMethod.SetMetricSamplingStrategy(RegMethod.RANDOM)
RegMethod.SetMetricSamplingPercentage(0.01)

RegMethod.SetInterpolator(sitk.sitkLinear)

""" Optimizer settings. """
RegMethod.SetOptimizerAsGradientDescent(learningRate=1.0, 
                                        numberOfIterations=500, 
                                        estimateLearningRate=RegMethod.Once)
RegMethod.SetOptimizerScalesFromPhysicalShift()

""" Setup for the multi-resolution framework. """     
RegMethod.SetShrinkFactorsPerLevel(shrinkFactors = [4,2,1])
RegMethod.SetSmoothingSigmasPerLevel(smoothingSigmas=[2,1,0])
RegMethod.SmoothingSigmasAreSpecifiedInPhysicalUnitsOn()

RegMethod.SetInitialTransform(InitialTx, inPlace=False)

FinalTx = RegMethod.Execute(FixIm, MovIm)

RegIm = sitk.Resample(SrcIm, TrgIm, FinalTx, sitk.sitkLinear, 0.0, SrcIm.GetPixelID())
```

Finally, I apply the final transform to `CubeIm`:

`TxCubeIm = sitk.Resample(CubeIm, TrgIm, FinalTx, sitk.sitkNearestNeighbor, 0.0, CubeIm.GetPixelID())`

Following is the resulting image:

![TxCube](https://user-images.githubusercontent.com/26715524/112360247-029c1d00-8cca-11eb-96c8-3fbea9fbb182.jpg)

As with the binary label of the ventricles posted in the SimpleElastix issue, the cube has been cropped.  A summary of the image attributes are as follows:

```
SrcSize = [512, 512, 192]
TrgSize = [270, 320, 40]

SrcSpacings = [0.5, 0.5, 1.0]
TrgSpacings = [0.71875, 0.71875, 4.000000562760051]

SrcDirs = [1.0, -2.051034e-10, 0.0, 2.051034e-10, 1.0, 0.0, -0.0, 0.0, 1.0]
TrgDirs = [0.97206213971436, 0.21732814336769, 0.08867736258088, 
           -0.1932394681254, 0.95541407584148, -0.2232542309624, 
           -0.13324302793231113, 0.19988101906970193, 0.9707182257087399]

SrcOrigin = [-129.80790963, -109.97568891, -117.140131]
TrgOrigin = [-63.550101835277, -156.40079761277, -57.175851875399]
```

I would greatly appreciate some insight as to where I'm going wrong in trying to resample a "source" label image to a "target" image grid based on a registration (or other) transform.  Thanks!

**Comment by zivy** [comment]
Hello @ctorti ,

Can you provide your input images and complete code so that we can run it and debug? It is hard to see what is going on just from the code.

**Comment by ctorti** [comment]
Hi @zivy ,

Thanks for your interest in helping.  

I wanted to export the 3D images as NIFTIs, as that would have allowed for performing the registration and resampling from the NIFTI imports, rather than having to run through the detailed processing required to get to the sitk images.  However, for unknown reasons, when I performed the registration using NIFTI-imported sitk images I got bizarre results.   

I checked the attributes of the NIFTI-imported sitk images (e.g. Origin, Directions, Spacing) and found there to be slight differences (it seems that floats are rounded to fewer significant digits when converting to NIFTI), but the differences looked too minor to explain the incorrect orientation the registered image (the orientation was not the same as the "fixed" image, nor the "moving" one).

Unfortunately as a result, I've had to include all of the code that leads from DICOMs to sitk images.  Here goes.

```
import SimpleITK as sitk
import numpy as np
from pydicom import dcmread
import os
import time
import matplotlib.pyplot as plt
from ipywidgets import interact, fixed
%matplotlib inline # for Anaconda Jupyter

def ImportDicomSeriesAsSitkImage(DicomDir, PixType='sitkFloat32'):
    """ Import a DICOM series as a SimpleITK image. """
    
    # Only a limited number of pixel types are catered for:
    PixTypesDict = {'sitkUInt8': sitk.sitkUInt8,
                    'sitkUInt16' : sitk.sitkUInt16,
                    'sitkUInt32' : sitk.sitkUInt32,
                    'sitkFloat32' : sitk.sitkFloat32,
                    'sitkFloat64' : sitk.sitkFloat64,
                    'sitkVectorUInt8' : sitk.sitkVectorUInt8,
                    'sitkVectorUInt16' : sitk.sitkVectorUInt16,
                    'sitkVectorFloat64' : sitk.sitkVectorFloat64}
    
    PixTypes = list(PixTypesDict.keys())
    
    if PixType in PixTypes:
        PixType = PixTypesDict[PixType]
    else:
        raise Exception("'PixType' '{PixType}' is not allowed.")
    
    
    Reader = sitk.ImageSeriesReader()
    Fnames = Reader.GetGDCMSeriesFileNames(DicomDir)
    Reader.SetFileNames(Fnames)
    
    Image = Reader.Execute()
    Image = sitk.Cast(Image, PixType)
    
    return Image


def GetDicomSOPuids(DicomDir):
    """ Get a list of SOP Instance UIDs for a DICOM series using SimpleITK's 
    ImageSeriesReader. """
    
    SeriesReader = sitk.ImageSeriesReader()
    
    FilePaths = SeriesReader.GetGDCMSeriesFileNames(DicomDir)
    
    FileReader = sitk.ImageFileReader()
    
    SOPuids = [] 
    
    for fpath in FilePaths:
        FileReader.SetFileName(fpath)

        FileReader.ReadImageInformation()
    
        SOPuids.append(FileReader.GetMetaData('0008|0018'))
        
    return SOPuids


def GetRSOPuidsInPFFGS(Seg):
    """ Get the list of ReferencedSOPInstanceUIDs in the 
    Per-FrameFunctionalGroupsSequence of a SEG. """
    
    RSOPuids = [] 
    
    sequences = Seg.PerFrameFunctionalGroupsSequence
    
    for sequence in sequences:
        uid = sequence.DerivationImageSequence[0]\
                      .SourceImageSequence[0]\
                      .ReferencedSOPInstanceUID
        
        RSOPuids.append(uid)
        
    return RSOPuids


def GetFrame2SliceInds(Seg, SOPuids):
    """ Get a list of the slice numbers that correspond to each  
    Per-FrameFunctionalGroupsSequence in a SEG. """
    
    # Get the list of ReferencedSOPInstanceUIDs:
    RSOPuids = GetRSOPuidsInPFFGS(Seg)
    
    F2Sinds = [] 
    
    for i in range(len(RSOPuids)):
        F2Sinds.append(SOPuids.index(RSOPuids[i]))
        
    return F2Sinds


def ImportLabIm(SegFpath, DcmDir):
    DcmIm = ImportDicomSeriesAsSitkImage(DcmDir)
    
    Seg = dcmread(SegFpath)
    
    PixArr = Seg.pixel_array
    
    SOPuids = GetDicomSOPuids(DcmDir)
    
    F2Sinds = GetFrame2SliceInds(Seg, SOPuids)
    
    # Number of frames, rows and columns:
    F, R, C = PixArr.shape
    
    # Number of slices:
    S = len(SOPuids)
    
    LabArr = np.zeros((S, R, C), dtype='uint')
    
    for i in range(F):
        s = F2Sinds[i]

        LabArr[s] += PixArr[i]
        
    LabIm = sitk.GetImageFromArray(LabArr)
    LabIm.SetOrigin(DcmIm.GetOrigin())
    LabIm.SetSpacing(DcmIm.GetSpacing())
    LabIm.SetDirection(DcmIm.GetDirection())
    
    return LabIm


def RegisterImages(FixIm, MovIm, FinalInterp):
    """ FinalInterp currently only accepts 'Linear' or 'NearestNeighbor'. """
    
    if FinalInterp == 'NearestNeighbor':
        Interpolator = sitk.sitkLinear
    elif FinalInterp == 'NearestNeighbor':
        Interpolator = sitk.sitkNearestNeighbor
    else:
        raise Exception("'FinalInterp' must either be 'Linear' or 'NearestNeighbor'.")
        
    
    InitialTx = sitk.CenteredTransformInitializer(FixIm, MovIm, 
                                                  sitk.AffineTransform(FixIm.GetDimension()), 
                                                  sitk.CenteredTransformInitializerFilter.MOMENTS)
    
    RegMethod = sitk.ImageRegistrationMethod()

    """ Similarity metric settings. """
    RegMethod.SetMetricAsMattesMutualInformation(numberOfHistogramBins=50)
    RegMethod.SetMetricSamplingStrategy(RegMethod.RANDOM)
    RegMethod.SetMetricSamplingPercentage(0.05)

    RegMethod.SetInterpolator(sitk.sitkLinear)

    """ Optimizer settings. """
    RegMethod.SetOptimizerAsGradientDescent(learningRate=1.0, 
                                            numberOfIterations=500, 
                                            estimateLearningRate=RegMethod.Once)
    RegMethod.SetOptimizerScalesFromPhysicalShift()

    """ Setup for the multi-resolution framework. """     
    RegMethod.SetShrinkFactorsPerLevel(shrinkFactors = [4,2,1])
    RegMethod.SetSmoothingSigmasPerLevel(smoothingSigmas=[2,1,0])
    RegMethod.SmoothingSigmasAreSpecifiedInPhysicalUnitsOn()

    #RegMethod.SetInitialTransform(InitialTx, inPlace=False)
    RegMethod.SetInitialTransform(InitialTx)

    FinalTx = RegMethod.Execute(FixIm, MovIm)

    RegIm = sitk.Resample(MovIm, FixIm, FinalTx, Interpolator, 0.0, MovIm.GetPixelID())
    
    return RegIm, FinalTx


def GetIndOfNearestSlice(Im0, Ind0, Im1, UseCentre=True):
    """ Get the index of the slice in Im1 that is nearest in z-position to the Ind0^th slice in Im0. 
    
    Returns:
    *******
    
    Ind1 : integer
        The slice index in Im1 whose z-position is nearest to Im0[Ind0].
    
    MinDiff : float
        The difference in z-position between Im0[Ind0] and Im1[Ind1] in mm.
    """
    
    # The z-position of Im0 at Ind0:
    z0 = Im0.TransformIndexToPhysicalPoint([0,0,Ind0])[2]
    
    R1, C1, S1 = Im1.GetSize()
    
    if UseCentre:
        i = R1//2
        j = C1//2
    else:
        i = 0
        j = 0
    
    # All z-positions in Im1: """
    Z1 = [Im1.TransformIndexToPhysicalPoint([i,j,k])[2] for k in range(S1)]
    
    Diffs = [abs(z0 - Z1[k]) for k in range(S1)]
    
    MinDiff = min(Diffs)
    
    Ind1 = Diffs.index(MinDiff)
    
    return Ind1, MinDiff


def Image2PixArr(Image):
    """ Convert a SimpleITK image to a Numpy pixel array and get the non-empty
    frame-to-slice indices. """
    
    Nda = sitk.GetArrayFromImage(Image)
    
    F2Sinds = []

    for i in range(Nda.shape[0]):
        Sum = np.amax(Nda[i])

        if Sum:
            F2Sinds.append(i)
    
    PixArr = np.zeros((len(F2Sinds), Nda.shape[1], Nda.shape[2]))
    
    for i in range(len(F2Sinds)):
        PixArr[i] = Nda[F2Sinds[i]]    
    
    return PixArr, F2Sinds


def PlotTwoImages(Title0, Im0, Ind0, Title1, Im1, Ind1):
    PixArr0 = sitk.GetArrayViewFromImage(Im0)[Ind0,:,:]
    PixArr1 = sitk.GetArrayViewFromImage(Im1)[Ind1,:,:]
    
    z0 = Im0.TransformIndexToPhysicalPoint([0,0,Ind0])[2]
    z1 = Im1.TransformIndexToPhysicalPoint([0,0,Ind1])[2]
    
    Title0 = f'{Title0}\nz = {round(z0, 2)} mm'
    Title1 = f'{Title1}\nz = {round(z1, 2)} mm'
    
    plt.subplots(1, 2, figsize=(15,8))
    
    plt.subplot(1, 2, 1)
    plt.imshow(PixArr0, cmap=plt.cm.Greys_r);
    plt.title(Title0)
    
    plt.subplot(1, 2, 2)
    plt.imshow(PixArr1, cmap=plt.cm.Greys_r);
    plt.title(Title1)
    
    plt.show()
    
    return


def PlotThreeImages(Title0, Im0, Ind0, Title1, Im1, Ind1, Title2, Im2, Ind2):
    PixArr0 = sitk.GetArrayViewFromImage(Im0)[Ind0,:,:]
    PixArr1 = sitk.GetArrayViewFromImage(Im1)[Ind1,:,:]
    PixArr2 = sitk.GetArrayViewFromImage(Im2)[Ind2,:,:]
    
    z0 = Im0.TransformIndexToPhysicalPoint([0,0,Ind0])[2]
    z1 = Im1.TransformIndexToPhysicalPoint([0,0,Ind1])[2]
    z2 = Im2.TransformIndexToPhysicalPoint([0,0,Ind2])[2]
    
    Title0 = f'{Title0}\nz = {round(z0, 2)} mm'
    Title1 = f'{Title1}\nz = {round(z1, 2)} mm'
    Title2 = f'{Title2}\nz = {round(z2, 2)} mm'
    
    plt.subplots(1, 3, figsize=(15,8))
    
    plt.subplot(1, 3, 1)
    plt.imshow(PixArr0, cmap=plt.cm.Greys_r);
    plt.title(Title0)
    
    plt.subplot(1, 3, 2)
    plt.imshow(PixArr1, cmap=plt.cm.Greys_r);
    plt.title(Title1)
    
    plt.subplot(1, 3, 3)
    plt.imshow(PixArr2, cmap=plt.cm.Greys_r);
    plt.title(Title2)
    
    plt.show()
    
    return


def PlotRegResults(FixIm, MovIm, RegIm, FixInd, FixTitle='Fixed image', 
                   MovTitle='Moving image', RegTitle='Registered image'):
    
    MovInd, zDiff = GetIndOfNearestSlice(FixIm, FixInd, MovIm, True)
    
    FixPixArr = sitk.GetArrayViewFromImage(FixIm)[FixInd,:,:]
    MovPixArr = sitk.GetArrayViewFromImage(MovIm)[MovInd,:,:]
    RegPixArr = sitk.GetArrayViewFromImage(RegIm)[FixInd,:,:]
    
    FixZ = FixIm.TransformIndexToPhysicalPoint([0,0,FixInd])[2]
    MovZ = MovIm.TransformIndexToPhysicalPoint([0,0,MovInd])[2]
    RegZ = RegIm.TransformIndexToPhysicalPoint([0,0,FixInd])[2]
    
    FixTitle = f'{FixTitle}\nz = {round(FixZ, 2)} mm'
    MovTitle = f'{MovTitle}\nz = {round(MovZ, 2)} mm'
    RegTitle = f'{RegTitle}\nz = {round(RegZ, 2)} mm'
        
    plt.subplots(2, 3, figsize=(18,12), dpi=80)
    
    alpha = 0.5
    
    BlendedIm = (1.0 - alpha)*FixIm[:,:,FixInd] + alpha*RegIm[:,:,FixInd]
    
    DiffIm = FixIm[:,:,FixInd] - RegIm[:,:,FixInd]
    
    plt.subplot(2, 3, 1)
    plt.imshow(MovPixArr, cmap=plt.cm.Greys_r);
    plt.title(MovTitle)
    plt.axis('off')
    
    plt.subplot(2, 3, 2)
    plt.imshow(FixPixArr, cmap=plt.cm.Greys_r);
    plt.title(FixTitle)
    plt.axis('off')
    
    plt.subplot(2, 3, 3)
    plt.imshow(RegPixArr, cmap=plt.cm.Greys_r);
    plt.title(RegTitle)
    plt.axis('off')
    
    plt.subplot(2, 3, 4)
    plt.axis('off')
    
    plt.subplot(2, 3, 5)
    plt.imshow(sitk.GetArrayViewFromImage(BlendedIm), cmap=plt.cm.Greys_r);
    plt.title('Blended image')
    plt.axis('off')
    
    plt.subplot(2, 3, 6)
    plt.imshow(sitk.GetArrayViewFromImage(DiffIm), cmap=plt.cm.Greys_r);
    plt.title('Difference image')
    plt.axis('off')
    
    plt.show()
    
    return
```

I've placed the DICOM-SEG and DICOM files (as well as the NIFTIs of the DICOMs and label image that were problematic) [here](https://1drv.ms/u/s!AsFwW2so8HlygnPMurI9Xo_--_fu?e=uNM609).

Modify the directory and file paths accordingly:

```
SrcDcmDir = r'C:\.....\2021-03-25_Cropped_segments\Ses4_Ser11_DICOMs'
TrgDcmDir = r'C:\.....\2021-03-25_Cropped_segments\Ses3_Ser6_DICOMs'

SrcSegFpath = r'C:\.....\2021-03-25_Cropped_segments\Ses4_Ser11_Ventricles_SEG_20210213_144855.dcm'

SrcDcmIm = ImportDicomSeriesAsSitkImage(SrcDcmDir)
TrgDcmIm = ImportDicomSeriesAsSitkImage(TrgDcmDir)

SrcLabIm = ImportLabIm(SrcSegFpath, SrcDcmDir)
```

The Source and Target DICOM images can be viewed as follows:

```
interact(PlotTwoImages,
         Title0='Source DICOM image',
         Im0=fixed(SrcDcmIm),
         Ind0=(0, SrcDcmIm.GetSize()[2]-1),
         Title1='Target DICOM image',
         Im1=fixed(TrgDcmIm),
         Ind1=(0, TrgDcmIm.GetSize()[2]-1));
```

![Source_and_Target_images](https://user-images.githubusercontent.com/26715524/112497994-04271d00-8d7e-11eb-9b42-ab48dfe1460b.jpg)

Next register `SrcDcmIm` to `TrgDcmIm`:

`RegDcmIm, FinalTx = RegisterImages(FixIm=TrgDcmIm, MovIm=SrcDcmIm, FinalInterp='NearestNeighbor')`

Display the result:

`PlotRegResults(TrgDcmIm, SrcDcmIm, RegDcmIm, 19)`

![Registration_result](https://user-images.githubusercontent.com/26715524/112497492-94b12d80-8d7d-11eb-882b-2d055bc34840.jpg)

```
interact(PlotThreeImages,
         Title0='Source DICOM image',
         Im0=fixed(SrcDcmIm),
         Ind0=(0, SrcDcmIm.GetSize()[2]-1),
         Title1='Target DICOM image',
         Im1=fixed(TrgDcmIm),
         Ind1=(0, TrgDcmIm.GetSize()[2]-1),
         Title2='Registered Source DICOM image',
         Im2=fixed(RegDcmIm),
         Ind2=(0, RegDcmIm.GetSize()[2]-1));
```

![Source_Target_and_Registered_images](https://user-images.githubusercontent.com/26715524/112498948-e7d7b000-8d7e-11eb-8252-8a385a680921.jpg)

Finally, resample `SrcLabIm` using `FinalTx`:

`TxSrcLabIm = sitk.Resample(SrcLabIm, TrgDcmIm, FinalTx, sitk.sitkNearestNeighbor, 0.0, SrcLabIm.GetPixelID())`

And display the result:

```
SrcPixArr, SrcLabF2Sinds = Image2PixArr(SrcLabIm)
TxSrcPixArr, TxSrcLabF2Sinds = Image2PixArr(TxSrcLabIm)

a = SrcLabF2Sinds[0]
b = SrcLabF2Sinds[-1]

c = TxSrcLabF2Sinds[0]
d = TxSrcLabF2Sinds[-1]

interact(PlotTwoImages,
         Title0='Source label image',
         Im0=fixed(SrcLabIm),
         Ind0=(a, b),
         Title1='Transformed label image',
         Im1=fixed(TxSrcLabIm),
         Ind1=(c, d));
```

![Source_and_Resampled_label_images](https://user-images.githubusercontent.com/26715524/112499472-6896ac00-8d7f-11eb-98ea-e6930dac27d0.jpg)

Hopefully things are clear enough to recreate the end result.  Thanks!

**Comment by zivy** [comment]
Hello @ctorti,
TLDR: 
Welcome to the wonderful world of 3D 😉 .  All is well with your results, it's the 2D visualization that is confusing you, due to the rotation required to map between the two images.

As your code is a bit too much, I wrote the following short script to work with the nifti images (registration isn't optimized and I can see that it isn't doing great, but it is good enough to illustrate the issue):
```
import SimpleITK as sitk

ses3_image = sitk.ReadImage('Ses3_Ser6_DcmIm.nii')
ses4_image = sitk.ReadImage('Ses4_Ser11_DcmIm.nii')
ses4_segmentation = sitk.ReadImage('Ses4_Ser11_Ventricles_LabIm.nii')

fixed_image = ses3_image
moving_image = ses4_image

transform = sitk.CenteredTransformInitializer(fixed_image, 
                                              moving_image, 
                                              sitk.AffineTransform(3), 
                                              sitk.CenteredTransformInitializerFilter.GEOMETRY)
registration_method = sitk.ImageRegistrationMethod()

# Similarity metric settings.
registration_method.SetMetricAsMattesMutualInformation(numberOfHistogramBins=50)
registration_method.SetMetricSamplingStrategy(registration_method.RANDOM)
registration_method.SetMetricSamplingPercentage(0.01)

registration_method.SetInterpolator(sitk.sitkLinear)

# Optimizer settings.
registration_method.SetOptimizerAsGradientDescent(learningRate=0.5, numberOfIterations=500, convergenceMinimumValue=1e-6, convergenceWindowSize=10)
registration_method.SetOptimizerScalesFromPhysicalShift()

registration_method.SetInitialTransform(transform, inPlace=True)
registration_method.Execute(sitk.Cast(fixed_image, sitk.sitkFloat32), 
                            sitk.Cast(moving_image, sitk.sitkFloat32))

ses3_segmentation = sitk.Resample(ses4_segmentation, ses3_image, transform = sitk.Transform(), interpolator = sitk.sitkNearestNeighbor)
sitk.WriteImage(ses3_segmentation, 'Ses3_Ser6_Ventricles_Resampled_Identity.nii')

ses3_segmentation = sitk.Resample(ses4_segmentation, ses3_image, transform = transform, interpolator = sitk.sitkNearestNeighbor)
sitk.WriteImage(ses3_segmentation, 'Ses3_Ser6_Ventricles_Resampled_Registered.nii')
```

Now we visualize the segmentation mapped onto the Ses3_Ser6_DcmIm.nii image before and after registration, using a viewer that does 3D (ITK-SNAP).

Before registration (resampled using identity transform):
<img width="963" alt="beforeRegistration" src="https://user-images.githubusercontent.com/338645/112653005-64709a00-8e24-11eb-98ea-2b8f97c843ed.png">

After registration (resampled using estimated transform):
<img width="966" alt="afterRegistration" src="https://user-images.githubusercontent.com/338645/112653072-75211000-8e24-11eb-8afe-1129e22303cf.png">

Hope this clarifies things.

**Comment by ctorti** [comment]
Hi @zivy ,

Thanks for your comments.  I ran the code you provided and recreated the same NIFTIs.  Indeed something seems to have gone wrong in the conversion from sitk Image to NIFTI so that the results don't look right.  It appears that despite only a minor* loss of significant figures the orientation (possibly other attributes too) is altered.  

*Perhaps the loss I consider to be minor is more significant than I appreciate.

Yet when I run the code on the DICOM series I get better registration results.  I've also noticed, somewhat surprisingly, that if I export that result as a NIFTI the orientation issue does not manifest - as evidenced by visualising in ITK-SNAP.

![Res_reg_lab_im_working_from_NIFTI](https://user-images.githubusercontent.com/26715524/112712651-cedb1600-8ec8-11eb-9da4-77743ca70b50.jpg)
Result working with NIFTIs

![Res_reg_lab_im_working_from_DICOMs](https://user-images.githubusercontent.com/26715524/112712680-dc909b80-8ec8-11eb-8678-c679dc43e8a9.jpg)
Result working with DICOMs

More importantly however, I'm afraid I don't understand your initial statement:

"All is well with your results, it's the 2D visualization that is confusing you, due to the rotation required to map between the two images."

If the Ventricles SEG was merely a (single-framed) segmentation, I would expect cropping to occur due to the different orientations of the moving and fixed image grids.  Likewise if the SEG was a (multi-framed) segment consisting of a sufficently small (given the slice thickness of the fixed image grid) number of frames, I would also expect cropping.

But in this case, the 3D volume is of sufficient height along the scan direction such that when "sliced" from the fixed image planes perspective, there should be full cross sections everywhere except near the boundaries.  How close to the boundaries will depend on the relative directions and sampling, as I've tried to work out below.

Following are comparisons of the moving and fixed image attributes (rounded to 3 significant figures):

Moving spacing:   (0.5, 0.5, 1.0)
Moving direction: (1.0, 0.000, 0.0, -0.000, 1.0, 0.0, 0.0, 0.0, 1.0)

Fixed size:       (270, 320, 40)
Fixed spacing:    (0.719, 0.719, 4.000)
Fixed direction:  (0.972, -0.193, -0.133, 0.217, 0.955, 0.120, 0.089, -0.223, 0.971)

There's a 0.24 deg angle between the z unit vectors.  Taking that into account, the effective distance between two fixed planes from the perspective of the moving grid is 4.1 mm.  So roughly speaking 4 moving frames should be resampled to 1 fixed frame. 

In a "worst-case" scenario, a single moving segmentation/frame would fall within a pair of fixed planes, resulting in a highly cropped appearance.  So too might 2 or 3 moving segmentations/frames, with diminishing degrees of cropping.  But I would expect any segmentation that is at least 4 slices away from either end to completely span the 3D volume along that cross section.   I don't see that here.   I hope I've explained that clearly.

Unless I'm looking at this the wrong way or misinterpreting what I'm seeing?
